### ISSUE TITLE
Unexpected function clause matching behavior

### ISSUE THREAD

**Issue by mdowell12** [elixir-lang/elixir#5686]
Hello!  I've been experiencing some strange behavior with function clause matching.  Included below is an example.

### Environment

* Elixir & Erlang versions:
     Elixir 1.3.4
     Erlang/OTP 19

Both were installed using the [asdf tool](https://github.com/asdf-vm/asdf).

* Operating system:
     macOS Sierra (version 10.12.1)

### Current behavior

In `example.exs`:

```elixir
    defmodule PatternMatchingBug do

      def my_function(:foo, "a") do
        IO.puts "inside first function"
      end

      def my_function(_, "b") do
        IO.puts "inside second function"
      end
    end

    PatternMatchingBug.my_function(:foo, "b")
```

In this case, my function call cannot find a matching definition:

    ** (FunctionClauseError) no function clause matching in PatternMatchingBug.my_function/2
        example.exs:4: PatternMatchingBug.my_function(:foo, "b")
        (elixir) lib/code.ex:363: Code.require_file/2

This alone is unexpected, given my understanding of pattern matching -- I thought that my second function was a valid match.  Further confusion comes when I add a "catchall" function, like in the following example:

```elixir
    defmodule PatternMatchingBug do

      def my_function(:foo, "a") do
        IO.puts "inside first function"
      end

      def my_function(_, "b") do
        IO.puts "inside second function"
      end

      def my_function(_, _) do
        IO.puts "inside catchall function"
      end
    end

    PatternMatchingBug.my_function(:foo, "b")
```

Now, my function matches as I expect, and I get the following output:

    inside second function

After all of this, I decided to retry my first example with a function guard, rather than using pattern matching:

```elixir
    defmodule PatternMatchingBug do

      def my_function(:foo, "a") do
        IO.puts "inside first function"
      end

      def my_function(_, arg) when arg == "b" do
        IO.puts "inside second function"
      end

    end

    PatternMatchingBug.my_function(:foo, "b")
```

This successfully matched on the second function, producing the following output:

    inside second function

Could you guys take a look?

Thanks!

**Comment by whatyouhide** [comment]
Hey @mdowell12, thanks for the report. Are you sure it was not a problem of, say, unsaved file or something similar? The first example works as expected in all the tests I did 😕 Can you try this in your shell:

```bash
cat << EOF > /tmp/clauses.ex
defmodule PatternMatchingBug do

  def my_function(:foo, "a") do
    IO.puts "inside first function"
  end

  def my_function(_, "b") do
    IO.puts "inside second function"
  end
end

PatternMatchingBug.my_function(:foo, "b")
EOF
elixir /tmp/clauses.ex
```

and let us know what's printed?

**Comment by josevalim** [comment]
Also please make sure that there is no `.beam` files in the current directory or of any warning that is printed while the program is executed.

**Comment by mdowell12** [comment]
@whatyouhide I tried your test above and still got the weird behavior:

    ** (FunctionClauseError) no function clause matching in PatternMatchingBug.my_function/2
        example.exs:4: PatternMatchingBug.my_function(:foo, "b")
        (elixir) lib/code.ex:363: Code.require_file/2

But then I switched up the elixir binary I was using.  I went from the binary I got from the [asdf-elixir plugin](https://github.com/asdf-vm/asdf-elixir) to a binary from this repo.  After doing so, I got the expected result:

    inside second function

It seems that the issue may be with the binary I got from asdf.  I'll close this and open an issue on their repo.  Thanks for the quick response and keep up the good work!
